File: combinatorics/resource/TMD/integer_partition/CODE/bai1/bai1.cpp
// problem1.cpp
// Full implementation for finding partitions of n into k parts
// and drawing their Ferrers and transposed Ferrers diagrams.

#include <iostream>
#include <vector>
#include <numeric>
#include <algorithm>
#include <string>

// Use the standard namespace for brevity in this small program.
using namespace std;

/**
 * @brief Draws the Ferrers diagram (F) for a given partition.
 * @param partition A vector of integers representing a partition.
 */
void draw_ferrers(const vector<int>& partition) {
    cout << "  Ferrers Diagram (F):" << endl;
    for (int part : partition) {
        cout << "    ";
        for (int i = 0; i < part; ++i) {
            cout << "* ";
        }
        cout << endl;
    }
}

/**
 * @brief Draws the transposed Ferrers diagram (F^T).
 * @param partition A vector of integers representing a partition.
 */
void draw_transpose(const vector<int>& partition) {
    cout << "  Transposed Ferrers Diagram (F^T):" << endl;
    if (partition.empty()) {
        return;
    }

    int max_part = partition[0];
    for (int i = 1; i <= max_part; ++i) {
        int count = 0;
        // The length of row i is the number of parts in the original partition >= i.
        for (int part : partition) {
            if (part >= i) {
                count++;
            }
        }
        cout << "    ";
        for (int j = 0; j < count; ++j) {
            cout << "* ";
        }
        cout << endl;
    }
}

/**
 * @brief Core recursive function using backtracking to find partitions.
 * @param target_sum The remaining sum to be partitioned.
 * @param num_parts The number of remaining parts to find.
 * @param max_val The maximum value for the next part (to ensure non-increasing order).
 * @param current_partition The partition being built in the current recursive path. Passed by reference.
 * @param results A vector to store all found partitions. Passed by reference.
 */
void find_partitions_recursive(int target_sum, int num_parts, int max_val,
                               vector<int>& current_partition,
                               vector<vector<int>>& results) {
    // Base Case: If only one part remains, it must equal the entire remaining sum.
    if (num_parts == 1) {
        if (target_sum >= 1 && target_sum <= max_val) {
            current_partition.push_back(target_sum);
            results.push_back(current_partition);
            current_partition.pop_back(); // Backtrack for the next exploration.
        }
        return;
    }

    // Recursive Step: Try all possible values for the current part.
    for (int i = max_val; i >= 1; --i) {
        int remaining_sum = target_sum - i;
        int remaining_parts = num_parts - 1;

        // Pruning: Check if a valid partition is still possible.
        if (remaining_sum >= remaining_parts && remaining_sum <= remaining_parts * i) {
            current_partition.push_back(i);
            find_partitions_recursive(remaining_sum, remaining_parts, i, current_partition, results);
            current_partition.pop_back(); // Backtrack after exploring this branch.
        }
    }
}

/**
 * @brief Wrapper function to initialize and start the recursive process.
 * @param n The integer to partition.
 * @param k The number of parts.
 * @return A vector of vectors, containing all valid partitions.
 */
vector<vector<int>> generate_all_partitions(int n, int k) {
    vector<vector<int>> results;
    if (n < k || k <= 0) {
        return results;
    }
    
    vector<int> current_partition;
    // The initial max_val is n - (k-1) because k-1 parts must be at least 1.
    int initial_max_val = n - (k - 1);
    find_partitions_recursive(n, k, initial_max_val, current_partition, results);
    return results;
}

/**
 * @brief Main function to drive the program.
 */
int main() {
    int n, k;
    cout << "--- Problem 1: Ferrers Diagrams & Partitions ---" << endl;
    cout << "Enter the integer to partition (n): ";
    cin >> n;
    cout << "Enter the number of parts (k): ";
    cin >> k;

    // Input validation
    if (cin.fail() || n <= 0 || k <= 0) {
        cerr << "Error: Invalid input. Please enter positive integers." << endl;
        return 1;
    }

    // Generate all partitions.
    auto partitions = generate_all_partitions(n, k);
    cout << string(50, '-') << endl;
    
    if (partitions.empty()) {
        cout << "There are no partitions of " << n << " into " << k << " parts (p_" << k << "(" << n << ") = 0)." << endl;
    } else {
        cout << "Found " << partitions.size() << " partitions of " << n << " into " << k << " parts (p_" << k << "(" << n << ") = " << partitions.size() << ")." << endl;
        
        for (size_t i = 0; i < partitions.size(); ++i) {
            // Print the partition itself
            cout << "\n[" << i + 1 << "] Partition: [";
            for (size_t j = 0; j < partitions[i].size(); ++j) {
                cout << partitions[i][j] << (j == partitions[i].size() - 1 ? "" : ", ");
            }
            cout << "]" << endl;

            // Draw the diagrams
            draw_ferrers(partitions[i]);
            draw_transpose(partitions[i]);
            cout << string(50, '-') << endl;
        }
    }

    return 0;
}

File: combinatorics/resource/TMD/integer_partition/CODE/bai1/bai1.py
# problem1.py
# Full implementation for finding partitions of n into k parts
# and drawing their Ferrers and transposed Ferrers diagrams.

import sys

def find_partitions_recursive(target_sum, num_parts, max_val, current_partition, results):
    """
    Core recursive function using backtracking to find partitions.

    Args:
        target_sum (int): The remaining sum to be partitioned.
        num_parts (int): The number of remaining parts to find.
        max_val (int): The maximum value for the next part (to ensure non-increasing order).
        current_partition (list): The list of parts found so far in the current recursive branch.
        results (list): A list to aggregate all valid partitions found.
    """
    # --- Base Case ---
    # If only one part remains, it must equal the entire remaining sum.
    if num_parts == 1:
        # Check if this final value is valid (must be >= 1 and <= max_val).
        if 1 <= target_sum <= max_val:
            # Add the final part and save the complete partition.
            results.append(current_partition + [target_sum])
        return

    # --- Recursive Step ---
    # Iterate through all possible values 'i' for the current part.
    # The loop goes downwards from max_val to ensure non-increasing order (λ_i >= λ_{i+1}).
    for i in range(max_val, 0, -1):
        remaining_sum = target_sum - i
        remaining_parts = num_parts - 1
        
        # --- Pruning to improve efficiency ---
        # Minimum condition: The remaining sum must be large enough for the remaining parts (at least 1 each).
        # Maximum condition: The remaining sum cannot be too large (at most 'i' for each remaining part).
        if remaining_sum >= remaining_parts and remaining_sum <= remaining_parts * i:
            find_partitions_recursive(
                remaining_sum, 
                remaining_parts, 
                i, # The max value for the next call is the current value 'i'.
                current_partition + [i], 
                results
            )

def generate_all_partitions(n, k):
    """
    Wrapper function to initialize and start the recursive process for finding partitions.
    
    Args:
        n (int): The integer to partition.
        k (int): The number of parts.

    Returns:
        list[list[int]]: A list of all partitions, where each partition is a list of integers.
    """
    if n < k or k <= 0:
        return []
    
    results = []
    # The initial max_val is n - (k-1) because the other k-1 parts must be at least 1.
    initial_max_val = n - (k - 1)
    find_partitions_recursive(n, k, initial_max_val, [], results)
    return results

def draw_ferrers(partition):
    """
    Draws the Ferrers diagram (F) for a given partition.
    
    Args:
        partition (list[int]): A list of integers representing a partition.
    """
    print("  Ferrers Diagram (F):")
    for part in partition:
        print(f"    {'* ' * part}")

def draw_transpose(partition):
    """
    Draws the transposed Ferrers diagram (F^T).
    
    Args:
        partition (list[int]): A list of integers representing a partition.
    """
    print("  Transposed Ferrers Diagram (F^T):")
    if not partition:
        return
        
    # The number of rows in the transposed diagram equals the largest part of the original partition.
    max_part = partition[0]
    
    # Iterate through each row of the transposed diagram.
    for i in range(1, max_part + 1):
        # The length of row i (λ'_i) is the number of parts in λ that are >= i.
        count = sum(1 for part in partition if part >= i)
        print(f"    {'* ' * count}")

def main():
    """
    Main function to drive the program. It gets user input, finds partitions,
    and prints the results along with their diagrams.
    """
    print("--- Problem 1: Ferrers Diagrams & Partitions ---")
    try:
        n = int(input("Enter the integer to partition (n): "))
        k = int(input("Enter the number of parts (k): "))
        if n <= 0 or k <= 0: 
            raise ValueError("n and k must be positive integers.")
    except ValueError as e:
        print(f"Error: Invalid input. {e}", file=sys.stderr)
        return

    # Generate all partitions of n into k parts.
    partitions = generate_all_partitions(n, k)
    print("-" * 50)
    
    if not partitions:
        print(f"There are no partitions of {n} into {k} parts (p_{k}({n}) = 0).")
    else:
        print(f"Found {len(partitions)} partitions of {n} into {k} parts (p_{k}({n}) = {len(partitions)}).")
        for i, p in enumerate(partitions):
            print(f"\n[{i+1}] Partition: {p}")
            draw_ferrers(p)
            draw_transpose(p)
            print("-" * 50)

if __name__ == "__main__":
    main()